### PR TITLE
Tag parsing

### LINKED ISSUE #210: Comments after tags are captured
If we have something like:

```gherkin
@tag#comment
Scenario: It does something
```

Then the tag will be something like `tag#comment`

Removing this line will produce a failing test
https://github.com/Behat/Gherkin/blob/master/tests/Behat/Gherkin/Cucumber/CompatibilityTest.php#L42

### PR DESCRIPTION
**Pull request by ciaranmcnulty**
Closes #210 and #213 

Not too happy with the implementation yet

For the tag whitespace we could optionally show a warning rather than an exception for backwards compatibility

### CODE REVIEW

File: src/Behat/Gherkin/Lexer.php [hunk 1]
```diff
@@ -502,6 +502,7 @@ protected function scanTableRow()
     protected function scanTags()
     {
         $line = $this->getTrimmedLine();
+        $line = preg_replace('/\s+#.*$/', '', $line);
```
Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Lexer.php:505]: Just ignoring comments feels wrong in the Lexer, but we don't have a consumeLineUpTo() - maybe that's a better approach so it'd be followed by a Comment with the same line number?

File: src/Behat/Gherkin/Parser.php [hunk 2]
```diff
@@ -635,6 +638,26 @@ protected function popTags()
...
+    /**
+     * Checks the tags fit the required format
+     *
+     * @param string[] $tags
+     */
+    protected function guardTags(array $tags)
+    {
+        foreach ($tags as $tag) {
+            if (preg_match('/\s/', $tag)) {
+                throw new ParserException(
```
Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Parser.php:650]: This could be `E_USER_DEPRECATED` if we think whitespace tags are a valid current usage rather than an error

**Comment by stof** [comment]
I would vote for adding such API to consume the line only partially instead of the hack about `#`

**Comment by ciaranmcnulty** [comment]
@stof Yeah I will add something like `consumeLineUntil($char)` or similar

**Comment by stof** [comment]
@ciaranmcnulty I suggest confirming upstream whether comments are meant to be allowed at the end of any line. If the answer is yes, it might be a matter of changing `consumeLine` instead.

**Comment by ciaranmcnulty** [comment]
It looks like they're only allowed at the end of tags lines, added in https://github.com/cucumber/cucumber/pull/880

File: src/Behat/Gherkin/Parser.php (hunk 2 above)
Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Parser.php:650]: What do you think about this @stof?

File: src/Behat/Gherkin/Lexer.php [hunk 3]
```diff
@@ -502,16 +511,24 @@ protected function scanTableRow()
...
+
         if (!isset($line[0]) || '@' !== $line[0]) {
             return null;
         }
 
+        if(preg_match('/^(?<line>.*)\s+#.*$/', $line, $matches)) {
+            ['line' => $line] = $matches;
+            $this->consumeLineUntil(mb_strlen($line, 'utf-8'));
+        }
+        else {
```
Comment (stof) [inline src/Behat/Gherkin/Lexer.php:523]: why don't you follow PSR-12 here ?

File: src/Behat/Gherkin/Parser.php (hunk 2 above)
Comment (stof) [reply src/Behat/Gherkin/Parser.php:650]: do tag expressions in TagFilter support tags with whitespace in them or no ? If such tags are not usable in TagFilter, then I would go for a ParserException

File: src/Behat/Gherkin/Lexer.php (hunk 3 above)
Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Lexer.php:523]: I don't find it very valuable, but I'll change

Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Lexer.php:523]: Hm maybe that sounded too rude. It's PSR-12 I dislike, your reviews are invaluable

Comment (stof) [reply src/Behat/Gherkin/Lexer.php:523]: I mean, the existing code already follows PSR-12 (or maybe the older PSR-2, we probably don't have code relying on new PHP features that were encoded in PSR-12). This is about being consistent.

**Comment by ciaranmcnulty** [comment]
TagFilter is quite happy to take / split by tags with whitespace so there's a chance there are people out there depending on this functionality.

trigger_error with a deprecation warning?

**Comment by stof** [comment]
@ciaranmcnulty then a deprecation is indeed better (and probably also in TagFilter if the filter uses a tag with whitespace)

**Comment by ciaranmcnulty** [comment]
I think #220 is necessary to test the deprecations properly

**Comment by ciaranmcnulty** [comment]
@stof would appreciate feedback on how the deprecation is implemented

**Comment by funkyproject** [comment]
Hello , 

That trigger a deprecation notice when we use a filter likes "@tag1 && @tag2". 
Even if the tags haven't spaces.

**Pull request merged**